Add ESLint config

pdehaan opened the pull request:
Fixes #118

pdehaan commented:
Still a bit WIPy. Future revisions would need to wire this fully into npm scripts and CI (and fix the errors so CI can pass).
I've just been testing this locally via <kbd>npx eslint .</kbd> (after running `npm install eslint eslint-plugin-import -D` locally).

pdehaan commented on .eslintrc.js:
```diff
@@ -0,0 +1,42 @@
+    sourceType: "module",
+  },
+  plugins: [
+    "import",
```
Ref: https://github.com/mozilla-ion/ion-core-addon/blob/43da9e64a31b42291256f3681642e30ade34feac/src/components/forwardEvents.js#L2

pdehaan commented on .eslintrc.js:
```diff
@@ -0,0 +1,46 @@
+        mocha: true,
+      },
+      extends: [
+        "plugin:mocha/recommended",
```
Added mocha rules to the tests/* directory since it flagged a few "no-async-describe" usages.

```
/Volumes/Dev/github/mozilla/ion-core-addon/tests/core-addon/Storage.test.js
   6:5   error  'sinon' is assigned a value but never used  no-unused-vars
  15:25  error  Unexpected async function in describe()     mocha/no-async-describe
  25:38  error  Unexpected async function in describe()     mocha/no-async-describe
  71:38  error  Unexpected async function in describe()     mocha/no-async-describe
```

Dexterp37 commented on package.json:
```diff
@@ -9,8 +9,11 @@
     "storybook": "start-storybook -p 6006",
-    "build-storybook": "build-storybook"
+    "build-storybook": "build-storybook",
+    "test": "npm run test-addon"
```
I slightly changed #27 , because we now have a confusing explosion of commands :)

Dexterp37 commented on package.json:
```diff
@@ -38,8 +47,23 @@
+  "engines": {
+    "node": ">=14.0.0"
+  },
+  "description": "This is the Core Add-on for Ion studies. It is implemented as a cross-browser WebExtension, using Svelte for the UI.",
```
```suggestion
  "description": "This is the Ion Core Add-on: a cross-browser WebExtension that allows users to participate to Ion studies.",
```

Dexterp37 commented on .eslintignore:
```diff
@@ -0,0 +1 @@
+public
```
Shouldn't this be `public/addon-build` and `public/build`? We also have other files in there (but ok, most of them don't require linting since they are images :) ).

Dexterp37 commented on .eslintrc.js:
```diff
@@ -0,0 +1,60 @@
+module.exports = {
```
nit: should we add the MPL license header here?

Dexterp37 reviewed (changes requested):
First of all, thank you @pdehaan for this contribution! I know this is WIP, but I gave it a first look anyway. In addition to the comments below, what about adding the CI step here, as well?

We just need to change this entry [here](https://github.com/mozilla-ion/ion-core-addon/blob/ce96118e92b988fc18b3f416c17f1284f34d181b/.circleci/config.yml#L52-L53) like this:

```yml
      - run:
          name: Run linting
          command: npm run lint
```

This would make sure we don't accidentally merge code that would break linting, after this PR is merged.

pdehaan commented:
Addressed feedback. CI is failing (as expected, until we/I try addressing the ESLint errors -- or I can fix the easy ones and downgrade the rest to warnings for somebody else to investigate/refactor).

Dexterp37 reviewed (approved):
Awesome, thanks for tackling this! I'm r+ (but let's *NOT* merge yet). I will be merging this tomorrow and addressing the follow-up work. Thank you for your contribution!

pdehaan commented:
Depending on how much we want to over-engineer this, we can also add the following potential rules (since I noticed some slight inconsistencies):

```js
    "eqeqeq": "warn",
    "no-unused-vars": [ "error", { vars: "all", args: "none", ignoreRestSiblings: false } ],
    "no-var": "off", // TODO: "warn",
    "prefer-const": "off", // TODO: "warn",
    "quotes": [ "warn", "double" ],
    "semi": [ "warn", "always" ],
```

I also found [**eslint-plugin-notice**](https://www.npmjs.com/package/eslint-plugin-notice) which might work nicely for enforcing MPL licenses in files.

```js
    "notice/notice": [ "error", { "mustMatch": "This Source Code Form is subject to the terms of the Mozilla Public" } ],
```

I can't figure out how to get all three lines of the MPL license header to match. I think it's a weird combination of whitespace and stray `*` chars, but I might be able to set the MPL template header in an external file. I also disabled the rule for .svelte files since it seems like none of the svelte files had MPL headers.

```js
    {
      files: ["*.svelte"],
      processor: "svelte3/svelte3",
      rules: {
        "notice/notice": "off",
      },
    },
```

Dexterp37 commented:
> 
> 
> Depending on how much we want to over-engineer this, we can also add the following potential rules (since I noticed some slight inconsistencies):

Thanks. I think that's probably for other issues/PRs though ;)
